kernel/src/expressions/scalars.rs

```diff
@@ -142,7 +142,7 @@ impl Scalar {
             Self::Decimal(_, precision, scale) => DataType::decimal_unchecked(*precision, *scale),
             Self::Null(data_type) => data_type.clone(),
             Self::Struct(data) => DataType::struct_type(data.fields.clone()),
-            Self::Array(data) => DataType::array_type(data.tpe.clone()),
+            Self::Array(data) => data.tpe.clone().into(),
```

Comment:
do we have the same `From` impl for the struct_type above?

Reply:
No, because it seemed one level of indirection too many to have `From<Iterator>`.

I _guess_ we could consider implementing [FromIterator](https://doc.rust-lang.org/std/iter/trait.FromIterator.html) for `StructType`? But we'd still need either `DataType::struct_type` or another `impl FromIterator`... and at that point things are uncomfortably indirect?
```rust
let data_type: DataType = some_iter.map(...).collect();
```

Also, there's no corresponding `TryFromIterator` trait to replace `StructType::try_new` and `DataType::try_struct_type`.

Reply:
yea makes sense. seems fine to leave :)

Reply:
Oh. My. Maybe we _should_ impl `FromIterator`. This actually works:
```rust
impl FromIterator<StructField> for StructType {
    fn from_iter<T: IntoIterator<Item = StructField>>(iter: T) -> Self {
        Self {
            type_name: "struct".into(),
            fields: iter.into_iter().map(|f| (f.name.clone(), f)).collect(),
        }
    }
}

impl StructType {
    pub fn new(fields: impl IntoIterator<Item = StructField>) -> Self {
        fields.into_iter().collect()
    }

    pub fn try_new<E>(fields: impl IntoIterator<Item = Result<StructField, E>>) -> Result<Self, E> {
        fields.into_iter().try_collect()
    }
```
(`Itertools::try_collect` Just Works)

But this would only be to make `new` and `try_new` nicer -- I checked all the various call sites and almost none of them benefit from using collect directly.

Reply:
I agree that's nice, but it'll be a mess if we ever need another argument for creating a struct, so given only local benefit I'd err on the side of _not_ implementing `FromIterator`